### ISSUE TITLE
kcl v0.10.10 is failing to start

### ISSUE THREAD

**Issue by fraenkel** [crossplane-contrib/function-kcl#206]
### What happened?
Upgrading to 0.10.10

```
{"level":"info","ts":1733853420.7700005,"caller":"fn/fn.go:37","msg":"Running Function","tag":""}
{"level":"info","ts":1733853420.7716818,"caller":"fn/fn.go:37","msg":"Running Function","tag":""}
{"level":"info","ts":1733853420.7729983,"caller":"fn/fn.go:37","msg":"Running Function","tag":""}
2024/12/10 17:57:00 failed to write file:  /tmp/kcl.mod open /tmp/kcl.mod: permission denied
```

### How can we reproduce it?
Just switched from 0.10.8 to 0.10.10

### What environment did it happen in?
Function version: 0.10.10

I see there was a previous issue which hit a similar issue over the KCL_CACHE_PATH.
The current image has `KCL_CACHE_PATH=/tmp`

**Comment by fraenkel** [comment]
So the v0.10.10 images have files in /tmp whereas they didn't before.

**Comment by vfarcic** [comment]
Hitting the same as well.

**Comment by vfarcic** [comment]
Can you remove the latest release until this is resolved? Otherwise, it'll fail for anyone using `>=` as the dependency version which results in the latest release of the function being deployed.

**Comment by zong-zhe** [comment]
> Can you remove the latest release until this is resolved? Otherwise, it'll fail for anyone using `>=` as the dependency version which results in the latest release of the function being deployed.

Sorry😞, I'm on my way home. This release will be removed soon when I get home.

**Comment by zong-zhe** [comment]
Hi @vfarcic, @fraenkel, @irizzant, @dmvariphy  

I've rolled back version 0.10.10 to a pre-release to prevent this issue from blocking your work.

**Comment by dmvariphy** [comment]
Thanks for doing that @zong-zhe!

**Comment by dzmoore** [comment]
Hey @zong-zhe, I'm still pulling `xpkg.upbound.io/crossplane-contrib/function-kcl:v0.10.10` with a `>=` dependency in my `crossplane.yaml` file (I think this is what @vfarcic was referring to in https://github.com/crossplane-contrib/function-kcl/issues/206#issuecomment-2541519942).  e.g.

```yaml
apiVersion: meta.pkg.crossplane.io/v1
kind: Configuration
# ...
spec:
  # ...
  dependsOn:
   # ...
    - function: xpkg.upbound.io/crossplane-contrib/function-kcl
      version: ">=v0.9.0"
```

```
$ k logs -n crossplane-system -l pkg.crossplane.io/function=crossplane-contrib-function-kcl
{"level":"info","ts":1734388399.8846216,"caller":"fn/fn.go:37","msg":"Running Function","tag":""}
2024/12/16 22:33:19 failed to write file:  /tmp/kcl.mod open /tmp/kcl.mod: permission denied

$ k get functions.pkg.crossplane.io crossplane-contrib-function-kcl 
NAME                              INSTALLED   HEALTHY   PACKAGE                                                    AGE
crossplane-contrib-function-kcl   True        False     xpkg.upbound.io/crossplane-contrib/function-kcl:v0.10.10   6m21s
```

Is there any chance that the package is still in the registry (maybe needs to be removed), or is there maybe a cache I need to clear locally?

I can work around it for now by manually changing `crossplane.yaml` and re-creating my Kind cluster, but just wanted to mention we're running into this.

**Comment by vfarcic** [comment]
Using `>=` as a dependency version is common in Crossplane meaning that many users of KCL functions are potentially affected. Please remove the latest release from the marketplace as a hot-fix.

**Comment by irizzant** [comment]
I agree, my renovate bot still shows 0.10.10 as available

**Comment by zong-zhe** [comment]
Hi @irizzant, @vfarcic, @dzmoore, @fraenkel, @dmvariphy

If you can, try the 0.10.10 release again to see if the problem persists, which may have been caused by some problems with the previously published process. I re-updated version 0.10.10 in the market

**Comment by vfarcic** [comment]
It works. Thanks a ton @zong-zhe.

**Comment by mproffitt** [comment]
Pulling `v0.10.10` has caused issues when using OCI images and I will also presume `github` due to missing CA certificates.

This is a regression of https://github.com/crossplane-contrib/function-kcl/issues/191 which was fixed in `0.10.10` and will affect some users.

Side question - is it worth pinning this issue for visibility until a new release is out that replaces `.10` or is a fix expected soon?

**Comment by zong-zhe** [comment]
> Pulling `v0.10.10` has caused issues when using OCI images and I will also presume `github` due to missing CA certificates.
> 
> This is a regression of #191 which was fixed in `0.10.10` and will affect some users.
> 
> Side question - is it worth pinning this issue for visibility until a new release is out that replaces `.10` or is a fix expected soon?

Hi @mproffitt 

We have found the issue and are working on fixing it for 0.10.10; soon, 0.10.10 will be re-released.

**Comment by zong-zhe** [comment]
Hi @fraenkel @irizzant @vfarcic @dmvariphy @dzmoore @mproffitt 😄 

I republished 0.10.10. If there is any problem, @ me here, thanks !

**Comment by Peefy** [comment]
This should be fixed upstream in [kcl-lang/kpm#566](https://github.com/kcl-lang/kpm/pull/566) — `feat: add a switch to control the auto-updating for mod file`. The PR added a flag that lets users control whether `kpm` auto-updates the `kcl.mod` file, which was the root cause of the unwanted write to `/tmp/kcl.mod` in v0.10.10.

The fix was merged on Dec 12, 2024 (commit `15d995b`). The issue is still open because function-kcl's image still ships with `KCL_CACHE_PATH=/tmp` and pre-existing `kcl.mod`/`kcl.mod.lock` files in that directory while running as a non-root user.

Recommended actions to fully resolve:

1. Bump `kcl-lang.io/kpm` in function-kcl's `go.mod` to a release that contains the auto-update switch (and use the switch to disable the unwanted auto-update).
2. Use a writable cache path in the function image (e.g., `/home/nonroot/.cache/kcl`) instead of `/tmp`, or run as a user that can write to `/tmp`.
3. Make sure the broken `v0.10.10` image tag is no longer resolvable in registries/marketplace (registry caching was also reported as a problem in the thread).

**Comment by Peefy** [comment]
Closing this — it was resolved back in December 2024 and has long since been overtaken by events.

### Recap

v0.10.10 shipped a base image that had `kcl.mod` / `kcl.mod.lock` baked into `/tmp` (@fraenkel's `ls /tmp` was the key observation). On start-up KCL tried to auto-update that mod file and hit `permission denied` on a read-only or differently-owned `/tmp`, so the function crash-looped. The reason it was intermittent — @dmvariphy seeing it under kind-on-WSL but not kind-on-Linux — is that whether the write succeeds depends on the `/tmp` permissions and uid mapping the runtime hands the container.

### Resolution

Three things happened, all of them complete:

1. **Immediate** — @zong-zhe rolled v0.10.10 back to a pre-release so `>=` dependency constraints stopped resolving to it, which is what @vfarcic, @dzmoore and @irizzant were blocked on.
2. **Re-release** — v0.10.10 was republished without the stray `/tmp` mod files. @vfarcic confirmed: *"It works. Thanks a ton @zong-zhe."*
3. **Root cause** — fixed upstream in kcl-lang/kpm#566, *"feat: add a switch to control the auto-updating for mod file"* (merged 2024-12-12), so KCL no longer unconditionally rewrites `kcl.mod` on load.

@mproffitt — the CA-certificate regression you raised (a re-break of #191) was part of the same botched image and went away with the republish. Please open a fresh issue if you still see it on a current release; it would be a different bug at this point.

### Current state

I checked `main`: there is no `KCL_CACHE_PATH=/tmp` and no committed `kcl.mod` anywhere in the repo or Dockerfile, so nothing reintroduces the original condition. The function is now on **v0.12.3** with `kpm` v0.12.9 / `lib` v0.12.5 — a completely different runtime generation from the one in this report.

Please reopen or file a new issue if anything resembling this shows up on a current version.
